packages/jest-cli/src/lib/watch_plugins_helpers.js

```diff
@@ -0,0 +1,51 @@
+/**
+ * Copyright (c) 2014-present, Facebook, Inc. All rights reserved.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ *
+ * @flow
+ */
+import type {GlobalConfig} from 'types/Config';
+import type {WatchPlugin, UsageData} from '../types';
+
+export const filterInteractivePlugins = (
+  watchPlugins: Array<WatchPlugin>,
+  globalConfig: GlobalConfig,
+): Array<WatchPlugin> => {
+  const usageInfos = watchPlugins.map(
+    p => p.getUsageInfo && p.getUsageInfo(globalConfig),
+  );
+
+  return watchPlugins.filter((plugin, i, array) => {
+    if (usageInfos[i]) {
+      const {key} = usageInfos[i];
+      return !usageInfos.slice(i + 1).some(u => u && key === u.key);
+    }
+
+    return false;
+  });
+};
+
+export const getSortedUsageRows = (
+  watchPlugins: Array<WatchPlugin>,
+  globalConfig: GlobalConfig,
+): Array<UsageData> => {
+  return filterInteractivePlugins(watchPlugins, globalConfig)
+    .sort((a: WatchPlugin, b: WatchPlugin) => {
+      if (a.isInternal) {
+        return -1;
+      }
+
+      const usageInfoA = a.getUsageInfo && a.getUsageInfo(globalConfig);
```

Comment:
Should the base class every watch plugin extends have a `getUsageInfo` function which returns `null` or empty string so you don't have to check for its existence?

Reply:
I think this will make the implementation easier, but I'm worried about making the API for the consumers a bit harder. Given that there will be plugins that do not implement it

Reply:
I sometimes wonder if we should just move everything to be a jest hook, and make the whole just around that.

```javascript
class MyPlugin {
  apply(jestHooks) {
    jestHooks.run(/* ... */);
    jestHooks.getUsageInfo(/* ... */);
    jestHooks.testRunComplete(/* ... */);
    jestHooks.fsChange(/* ... */);
    jestHooks.onData(/* ... */);
    jestHooks.shouldRunTestSuite(/* ... */);
  }
}
```
instead of 
```javascript
class MyPlugin {
  apply(jestHooks) {
    jestHooks.testRunComplete(/* ... */);
    jestHooks.fsChange(/* ... */);
    jestHooks.shouldRunTestSuite(/* ... */);
  }

  onData(/* ... */) {
    /* ... */
  }

  run(/* ... */) {
    /* ... */
  }

  getUsageInfo(/* ... */) {
    /* ... */
  }
}
```

Seems like the plugin architecture is working already, which means that changing to an API like the one above should not be that hard.

This means that the `apply` method is the only method that exists in the class and everything else is a jestHook. Seems that if we would like to make this changes it is better to do them now than later. 

thoughts? Which API seems cleaner? Which API would be easier to document?
@rickhanlonii @orta @cpojer @SimenB

Reply:
I'm worried about the first API being too open. There are other open source tools out there that allow people to do anything they want in a plugin and it generally causes the entire system to slow down considerably. I think we should stick with a more limited API that guides user to build plugins in the way we expect them to be built. Of course, it's still possible to completely break out of that system and do harmful slow things but it is less likely to happen.

Reply:
+1 to @cpojer   I like that the current Watch Plugin API is similar in structure to the runner and reporter - implement a class with a specific structure

Reply:
+1 - slow and deliberate improvements as we need them, I think this looks great

Reply:
Perfect! 😄